File: frontend/src/services/cartService.js
import api from "./api";

export const getCart = () => {

    return api.get("cart/");

};

export const addToCart = (

    productId,

    quantity

)=>{

    return api.post(

        "cart/add/",

        {

            product_id: productId,

            quantity,

        }

    );

};

export const updateCartItem=(

    id,

    quantity

)=>{

    return api.put(

        `cart/item/${id}/`,

        {

            quantity,

        }

    );

};

export const removeCartItem=(id)=>{

    return api.delete(

        `cart/item/${id}/delete/`

    );

};

File: backend/apps/cart/services/cart_service.py
from apps.cart.models import Cart, CartItem
from apps.products.models import Product


class CartService:

    @staticmethod
    def get_cart(user):
        cart, _ = Cart.objects.get_or_create(user=user)
        return cart

    @staticmethod
    def add_item(user, product_id, quantity):

        cart = CartService.get_cart(user)

        product = Product.objects.get(id=product_id)

        item, created = CartItem.objects.get_or_create(
            cart=cart,
            product=product,
            defaults={"quantity": quantity},
        )

        if not created:
            item.quantity += quantity

        if item.quantity > product.stock:
            raise ValueError(
                "Requested quantity exceeds available stock."
            )

        item.save()

        return cart

    @staticmethod
    def update_quantity(item, quantity):

        if quantity > item.product.stock:
            raise ValueError(
                "Insufficient stock."
            )

        item.quantity = quantity
        item.save()

        return item

    @staticmethod
    def remove_item(item):
        item.delete()